spring-kafka/src/main/java/org/springframework/kafka/support/DefaultKafkaHeaderMapper.java

```diff
@@ -266,30 +267,18 @@ public void fromHeaders(MessageHeaders headers, Headers target) {
 		final ObjectMapper headerObjectMapper = getObjectMapper();
 		headers.forEach((key, rawValue) -> {
 			if (matches(key, rawValue)) {
-				Object valueToAdd = headerValueToAddOut(key, rawValue);
-				if (valueToAdd instanceof byte[]) {
-					target.add(new RecordHeader(key, (byte[]) valueToAdd));
-				}
-				else {
-					try {
-						String className = valueToAdd.getClass().getName();
-						boolean encodeToJson = this.encodeStrings;
-						if (this.toStringClasses.contains(className)) {
-							valueToAdd = valueToAdd.toString();
-							className = JAVA_LANG_STRING;
-							encodeToJson = true;
-						}
-						if (!encodeToJson && valueToAdd instanceof String) {
-							target.add(new RecordHeader(key, ((String) valueToAdd).getBytes(getCharset())));
-						}
-						else {
-							target.add(new RecordHeader(key, headerObjectMapper.writeValueAsBytes(valueToAdd)));
-						}
-						jsonHeaders.put(key, className);
+				if (doesMatchMultiValueHeader(key)) {
+					Iterable<?> valuesToMap;
+					if (rawValue instanceof Iterable<?> iterable) {
+						valuesToMap = iterable;
 					}
-					catch (Exception e) {
-						logger.error(e, () -> "Could not map " + key + " with type " + rawValue.getClass().getName());
+					else {
+						valuesToMap = List.of(rawValue);
```

Comment:
Now, since we are very close to the final solution, let's talk about optimization!
This `fromHaders` is very active action performed on each message to produce to Kafka.
Therefore, any extra object is performance degradation.
I would prefer to not have this `List.of()` for a single value.
Even if `if..else` would have some duplication, it is still would be better by performance.
So, let' do `forEach` if `Iterable`, and just one for a single-value header!